lib/modules/mk*Module: give resulting modules meaningful keys

sternenseemann opened the pull request:
This enables us to disable unwanted modules created in this way using `disabledModules`. This is especially useful, since it is currently impossible to disable `rename.nix` (#245265), so it is required to get rid of the offending modules individually.

Undocumented until the bike shedding about the `key` format has finished…

###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

sternenseemann committed:
lib/modules/mk*Module: give resulting modules meaningful keys

This enables us to disable unwanted modules created in this way using
`disabledModules`. This is especially useful, since it is currently
impossible to disable `rename.nix` (#245265), so it is required to get
rid of the offending modules individually.

ncfavier commented on lib/modules.nix:
```diff
@@ -1034,7 +1034,8 @@ let
   mkRemovedOptionModule = optionName: replacementInstructions:
     { options, ... }:
-    { options = setAttrByPath optionName (mkOption {
+    { key = "lib/modules.nix#mkRemovedOptionModule-${showOption optionName}";
```
```suggestion
    { _file = "lib/modules.nix"; key = "mkRemovedOptionModule-${showOption optionName}";
```

Probably?

sternenseemann replied:
It seemed to me that `key` defaults to `_file` normally, so I assumed that it's not possible to pass `_file` and `key` to `disabledModules` to get more precise filtering, but I'd need to confirm that suspicion.

ncfavier replied:
I don't think that's possible, but I think `mkRemovedOptionModule-` is precise enough, and `_file` might help with error messages (?)

roberth replied:
`_file` is for error messages.

By setting `_file` we would hide the importing module's `_file`, which is the `_file` we actually want in error messages.
Usually we have
```nix
# foo.nix
{
  imports = [ (mkRemovedOptionModule [ "bar" ]) ];
}
```

and then `error in foo.nix` is more useful than `error in lib/modules.nix` or `error in removed option bar`.

So we should only set `key`.

infinisil reviewed (changes requested):
Needs some tests

roberth commented:
> some tests

the right number of tests.

sternenseemann committed:
lib/tests: move dummy warnings implementation into common module

A few tests use lib.mk*OptionModule which sometimes set config.warnings,
so the corresponding option needs to be defined. This commit moves this
into dummy-warnings-module.nix which can be reused in many tests.

sternenseemann commented:
Added tests for the three types of modules (doRename, mkRemovedOptionModule and mkMergedOptionModule). Are we happy with the name format the `key`s get? If yes, I'd also document them.

I realise that there is the possibility of having the same `key` for different modules with how the `key`s are currently formed. We could make this less likely to happen on accident by using more unusual separators than `-`.

infinisil commented:
I think we should figure out the root problem in https://github.com/NixOS/nixpkgs/issues/245265 first. If that can be fixed there should be no need for this PR (unless it has another use case other than working around #245265)
